lib/utils.js

```diff
@@ -0,0 +1,67 @@
+"use strict";
+
+function resolveHttpError(response, message) {
+    if (response.statusCode < 400) {
+        return;
+    }
+
+    message = message
+        ? message + " "
+        : "";
+
+    message += "Status code " + response.statusCode + " (" + response.statusMessage + ").";
+
+    if (response.body) {
+        message += " " + response.body;
+    }
+
+    var error = new Error(message);
+    error.response = response;
+
+    return error;
+}
+
+function createCallback(action, cb) {
+    return function (err, response) {
+        err = err || resolveHttpError(response, "Error " + action + ".");
+
+        if (err) {
+            return cb(err);
+        }
+
+        cb(null, response);
+    };
+}
+
+function createBodyCallback(action, cb) {
+    return createCallback(action, function (err, response) {
+        if (err) {
+            return cb(err);
+        }
+
+        cb(null, response.body, response);
+    });
+}
+
+function createJsonCallback(action, cb) {
+    return createCallback(action, function (err, response) {
+        if (err) {
+            return cb(err);
+        }
+
+        try {
+            var data = JSON.parse(response.body);
```

Comment:
I wonder now whether `request` should do all JSON parsing. It does it implicitly when a JSON body is sent, but it might be more consistent to set `json: true` and bring it down to two callback factories.